File: claude_exp/routers/router_py/router/dispatcher.py
import logging
import queue
import threading
import time

import iso8583

from router.trace import TraceRecorder
from router.upstream import write_upstream
from shared.ims_connect import build_frame

logger = logging.getLogger(__name__)

_STAN_MODULUS = 1_000_000
_RESPONSE_MTIS = ("0110", "0130", "0430")


class PendingEntry:
    __slots__ = ("up_conn", "up_write_lock", "upstream_stan", "created_at", "started_at")

    def __init__(self, up_conn, up_write_lock, upstream_stan, created_at, started_at=None):
        self.up_conn = up_conn
        self.up_write_lock = up_write_lock
        self.upstream_stan = upstream_stan
        self.created_at = created_at
        # Distinct from created_at (set once this entry is added to _pending, i.e. after crypto
        # + right before the downstream send - used for the TTL reaper and downstream_rtt).
        # started_at is the true transaction start (RoutedMessage.enqueued_at, before queue wait
        # and the upstream-leg crypto call) - used for the end-to-end "total" latency bucket.
        self.started_at = started_at if started_at is not None else created_at


class RoutedMessage:
    __slots__ = ("req", "up_conn", "up_write_lock", "up_addr", "raw", "enqueued_at")

    def __init__(self, req, up_conn, up_write_lock, up_addr, raw=b"", enqueued_at=None):
        self.req = req
        self.up_conn = up_conn
        self.up_write_lock = up_write_lock
        self.up_addr = up_addr
        self.raw = raw
        self.enqueued_at = enqueued_at


class Dispatcher:
    """Worker pool. Routes 0100 upstream -> crypto -> downstream.
    Routes 0110/0130/0430 downstream -> upstream (STAN lookup)."""

    def __init__(self, cfg, downstream, crypto, spec, stats, reconnect_event, upstream_spec=None):
        self.cfg = cfg
        self.downstream = downstream
        self.crypto = crypto
        self.spec = spec
        # Upstream-facing leg can speak a different ISO 8583 encoding than the downstream leg
        # (e.g. router_2's EBCDIC partner spec vs the shared downstream_host's ASCII spec) -
        # defaults to `spec` so single-spec configs (router_1, and every config predating this)
        # behave exactly as before.
        self.upstream_spec = upstream_spec if upstream_spec is not None else spec
        self.stats = stats
        self.reconnect_event = reconnect_event

        self.trace = TraceRecorder()

        self._queue = queue.Queue(maxsize=cfg.queue_maxsize)
        self._response_queue = queue.Queue(maxsize=cfg.queue_maxsize)
        self._pending = {}
        self._pending_lock = threading.Lock()
        self._stan_counter = 0
        self._stan_lock = threading.Lock()
        self._stop_event = threading.Event()
        self._worker_threads = []
        self._response_worker_threads = []
        self._reaper_thread = None

    def _next_stan(self) -> str:
        with self._stan_lock:
            self._stan_counter = (self._stan_counter + 1) % _STAN_MODULUS
            return str(self._stan_counter).zfill(6)

    def start(self) -> None:
        for i in range(self.cfg.worker_threads):
            t = threading.Thread(target=self._worker_loop, name=f"worker-{i}", daemon=True)
            t.start()
            self._worker_threads.append(t)
        for i in range(self.cfg.response_worker_threads):
            t = threading.Thread(
                target=self._response_worker_loop, name=f"response-worker-{i}", daemon=True
            )
            t.start()
            self._response_worker_threads.append(t)
        self._reaper_thread = threading.Thread(
            target=self._pending_reaper, name="pending-reaper", daemon=True
        )
        self._reaper_thread.start()

    def submit(self, msg: RoutedMessage) -> None:
        msg.enqueued_at = time.monotonic()
        self._queue.put(msg)
        qsize = self._queue.qsize()
        self.stats.set_gauge("queue_depth", qsize)
        logger.debug("dispatcher: queued mti=%s (queue_depth=%d)", msg.req.get("t"), qsize)

    def submit_response(self, resp: dict, raw: bytes = b"") -> None:
        """Enqueues a downstream response (0110/0130/0430) for handling by the response
        worker pool, instead of processing it inline on the caller's thread. This keeps the
        0110 leg's crypto call (validate_0110) from being a single-threaded bottleneck that's
        entirely separate from - and doesn't have to compete with - the 0100 leg's queue."""
        self._response_queue.put((resp, raw))
        qsize = self._response_queue.qsize()
        self.stats.set_gauge("response_queue_depth", qsize)
        logger.debug("dispatcher: queued response mti=%s (response_queue_depth=%d)", resp.get("t"), qsize)

    def _worker_loop(self) -> None:
        while True:
            msg = self._queue.get()
            self.stats.set_gauge("queue_depth", self._queue.qsize())
            if msg is None:
                return
            if msg.enqueued_at is not None:
                self.stats.record_latency("queue_wait", (time.monotonic() - msg.enqueued_at) * 1000)
            try:
                self._process(msg)
            except OSError:
                logger.warning("downstream send failed while dispatching; triggering reconnect")
                self.reconnect_event.set()
            except Exception:
                logger.exception("unexpected error processing dispatched message")

    def _response_worker_loop(self) -> None:
        while True:
            item = self._response_queue.get()
            self.stats.set_gauge("response_queue_depth", self._response_queue.qsize())
            if item is None:
                return
            resp, raw = item
            try:
                self.handle_response(resp, raw=raw)
            except Exception:
                logger.exception("unexpected error processing dispatched response")

    def _process(self, msg: RoutedMessage) -> None:
        req = msg.req
        mti = req.get("t")
        pan = req.get("2", "")
        upstream_stan = req.get("11", "")

        router_stan = self._next_stan()
        tracing = self.trace.start(router_stan, upstream_stan, pan, mti, msg.raw)

        fwd = dict(req)
        if mti == "0100":
            crypto_start = time.monotonic()
            result = self.crypto.validate("validate_0100", pan, req.get("47", ""), router_stan=router_stan)
            crypto_ms = (time.monotonic() - crypto_start) * 1000
            self.stats.record_latency("crypto_rtt", crypto_ms)
            if tracing:
                self.trace.hop(router_stan, "crypto_call", crypto_ms=round(crypto_ms, 3), enriched=bool(result))
            if result:
                fwd["47"] = result
        fwd["11"] = router_stan

        encoded, _ = iso8583.encode(fwd, self.spec)
        if tracing:
            self.trace.hop(router_stan, "downstream_send", raw=bytes(encoded))

        with self._pending_lock:
            if router_stan in self._pending:
                logger.error(
                    "router_stan %s still outstanding; overwriting pending entry",
                    router_stan, extra={"router_stan": router_stan},
                )
            self._pending[router_stan] = PendingEntry(
                up_conn=msg.up_conn,
                up_write_lock=msg.up_write_lock,
                upstream_stan=upstream_stan,
                created_at=time.monotonic(),
                started_at=msg.enqueued_at,
            )
            pending_count = len(self._pending)
        self.stats.set_gauge("pending_count", pending_count)

        frame = build_frame(
            0x00, self.cfg.downstream.irm_id, self.cfg.downstream.client_id, mti=fwd["t"], data=bytes(encoded)
        )
        self.downstream.send(frame)
        self.stats.record_sent()
        logger.debug(
            "dispatcher: forwarded mti=%s to downstream, upstream_stan=%s router_stan=%s",
            mti, upstream_stan, router_stan, extra={"router_stan": router_stan},
        )

    def handle_response(self, resp: dict, raw: bytes = b"") -> None:
        mti = resp.get("t")
        if mti == "0810":
            return
        if mti not in _RESPONSE_MTIS:
            logger.warning("unexpected response MTI from downstream: %s", mti)
            return

        router_stan = resp.get("11", "")
        self.trace.hop(router_stan, "downstream_recv", raw=raw)
        tracing = self.trace.is_tracing(router_stan)

        now = time.monotonic()
        with self._pending_lock:
            entry = self._pending.pop(router_stan, None)
            pending_count = len(self._pending)
        self.stats.set_gauge("pending_count", pending_count)
        if entry is None:
            logger.warning(
                "no pending entry for router_stan %s", router_stan, extra={"router_stan": router_stan}
            )
            if tracing:
                self.trace.finish(router_stan)
            return
        self.stats.record_latency("downstream_rtt", (now - entry.created_at) * 1000)

        fwd = dict(resp)
        fwd["11"] = entry.upstream_stan
        if mti == "0110":
            pan = resp.get("2", "")
            crypto_start = time.monotonic()
            result = self.crypto.validate("validate_0110", pan, resp.get("47", ""), router_stan=router_stan)
            crypto_ms = (time.monotonic() - crypto_start) * 1000
            self.stats.record_latency("crypto_rtt", crypto_ms)
            if tracing:
                self.trace.hop(router_stan, "crypto_call", crypto_ms=round(crypto_ms, 3), enriched=bool(result))
            if result:
                fwd["47"] = result

        encoded, _ = iso8583.encode(fwd, self.upstream_spec)
        if tracing:
            self.trace.hop(router_stan, "upstream_send", raw=bytes(encoded))
        try:
            with entry.up_write_lock:
                write_upstream(entry.up_conn, bytes(encoded), self.cfg.upstream)
            self.stats.record_sent()
            logger.debug(
                "dispatcher: forwarded mti=%s to upstream, router_stan=%s upstream_stan=%s",
                mti, router_stan, entry.upstream_stan, extra={"router_stan": router_stan},
            )
        except OSError:
            # entry.up_conn can be closed by session teardown racing this write.
            logger.warning(
                "failed to write response upstream for stan %s",
                entry.upstream_stan, extra={"router_stan": router_stan},
            )
        self.stats.record_latency("total", (time.monotonic() - entry.started_at) * 1000)
        if tracing:
            self.trace.finish(router_stan)

    def _pending_reaper(self) -> None:
        while not self._stop_event.wait(1.0):
            now = time.monotonic()
            expired = []
            with self._pending_lock:
                for stan, entry in list(self._pending.items()):
                    if now - entry.created_at > self.cfg.pending_ttl_seconds:
                        expired.append((stan, entry))
                for stan, _entry in expired:
                    del self._pending[stan]
                pending_count = len(self._pending)
            if expired:
                self.stats.set_gauge("pending_count", pending_count)

            for stan, entry in expired:
                logger.warning(
                    "pending entry %s expired after %ds; sending local decline",
                    stan, self.cfg.pending_ttl_seconds, extra={"router_stan": stan},
                )
                decline = {"t": "0110", "11": entry.upstream_stan, "39": "91"}
                try:
                    encoded, _ = iso8583.encode(decline, self.upstream_spec)
                    with entry.up_write_lock:
                        write_upstream(entry.up_conn, bytes(encoded), self.cfg.upstream)
                    self.stats.record_sent()
                except OSError:
                    logger.warning(
                        "failed to write expiry decline for stan %s",
                        entry.upstream_stan, extra={"router_stan": stan},
                    )

    def pending_snapshot(self) -> list:
        """Read-only view of in-flight transactions (downstream request sent, no response yet)
        for live diagnosis - e.g. "why is this router stuck", without waiting for the
        pending_ttl_seconds reaper to expire and decline them. Sorted oldest-first so a stuck
        transaction sorts to the top."""
        now = time.monotonic()
        with self._pending_lock:
            entries = [
                {
                    "router_stan": stan,
                    "upstream_stan": entry.upstream_stan,
                    "age_seconds": round(now - entry.created_at, 3),
                }
                for stan, entry in self._pending.items()
            ]
        entries.sort(key=lambda e: e["age_seconds"], reverse=True)
        return entries

    def purge(self) -> dict:
        dropped_queue = 0
        while True:
            try:
                self._queue.get_nowait()
                dropped_queue += 1
            except queue.Empty:
                break
        dropped_response_queue = 0
        while True:
            try:
                self._response_queue.get_nowait()
                dropped_response_queue += 1
            except queue.Empty:
                break
        with self._pending_lock:
            dropped_pending = len(self._pending)
            self._pending.clear()
        self.stats.set_gauge("queue_depth", self._queue.qsize())
        self.stats.set_gauge("response_queue_depth", self._response_queue.qsize())
        self.stats.set_gauge("pending_count", 0)
        return {
            "dropped_queue": dropped_queue,
            "dropped_response_queue": dropped_response_queue,
            "dropped_pending": dropped_pending,
        }

    def drain_and_stop(self) -> None:
        self._stop_event.set()
        for _ in self._worker_threads:
            self._queue.put(None)
        for _ in self._response_worker_threads:
            self._response_queue.put(None)
        for t in self._worker_threads:
            t.join(timeout=5)
        for t in self._response_worker_threads:
            t.join(timeout=5)
        if self._reaper_thread is not None:
            self._reaper_thread.join(timeout=5)

        # Session teardown (upstream or downstream disconnect, reconnect) previously discarded
        # any still-in-flight transactions with zero trace - unlike purge(), which reports
        # dropped_pending, this path left no log line explaining why a transaction just
        # vanished. There's nothing to *do* about them (the upstream client that would receive
        # a decline is already gone), but a live-diagnosis session needs the record.
        with self._pending_lock:
            dropped = list(self._pending.items())
            self._pending.clear()
        for stan, entry in dropped:
            logger.warning(
                "session torn down with router_stan %s still pending (upstream_stan=%s); abandoning",
                stan, entry.upstream_stan, extra={"router_stan": stan},
            )
        if dropped:
            logger.warning("session teardown abandoned %d pending transaction(s)", len(dropped))
            self.stats.set_gauge("pending_count", 0)

        abandoned_traces = self.trace.abandon_in_progress()
        if abandoned_traces:
            logger.warning("session teardown left %d in-progress trace(s) incomplete", len(abandoned_traces))

File: claude_exp/routers/router_cpp/src/router/dispatcher.cpp
#include "router/dispatcher.h"

#include <algorithm>
#include <cmath>
#include <cstdio>

#include "shared/framing.h"
#include "shared/ims_connect.h"
#include "shared/iso_codec.h"
#include "shared/log.h"

namespace router {

Dispatcher::Dispatcher(const RouterConfig& cfg, DownstreamConnection& downstream, CryptoClient& crypto,
                         shared::Stats& stats, shared::StopEvent& reconnect_event)
    : cfg_(cfg), downstream_(downstream), crypto_(crypto), stats_(stats), reconnect_event_(reconnect_event) {}

Dispatcher::~Dispatcher() { drain_and_stop(); }

void Dispatcher::start() {
    for (int i = 0; i < cfg_.worker_threads; ++i) {
        workers_.emplace_back([this] { worker_loop(); });
    }
    for (int i = 0; i < cfg_.response_worker_threads; ++i) {
        response_workers_.emplace_back([this] { response_worker_loop(); });
    }
    reaper_ = std::thread([this] { reaper_loop(); });
}

std::string Dispatcher::next_stan() {
    int next = (stan_counter_.fetch_add(1) + 1) % 1'000'000;
    char buf[8];
    std::snprintf(buf, sizeof(buf), "%06d", next);
    return std::string(buf);
}

void Dispatcher::submit(RoutedMessage msg) {
    std::string mti = msg.req.count("t") ? msg.req.at("t") : "";
    msg.enqueued_at = std::chrono::steady_clock::now();
    std::unique_lock<std::mutex> lock(queue_mutex_);
    queue_cv_not_full_.wait(lock, [this] { return queue_.size() < static_cast<size_t>(cfg_.queue_maxsize); });
    queue_.push_back(std::move(msg));
    stats_.set_gauge("queue_depth", static_cast<int>(queue_.size()));
    queue_cv_not_empty_.notify_one();
    LOG_DEBUG("dispatcher: queued mti=" + mti + " (queue_depth=" + std::to_string(queue_.size()) + ")");
}

void Dispatcher::worker_loop() {
    while (true) {
        std::optional<RoutedMessage> item;
        {
            std::unique_lock<std::mutex> lock(queue_mutex_);
            queue_cv_not_empty_.wait(lock, [this] { return !queue_.empty(); });
            item = std::move(queue_.front());
            queue_.pop_front();
            stats_.set_gauge("queue_depth", static_cast<int>(queue_.size()));
            queue_cv_not_full_.notify_one();
        }
        if (!item.has_value()) break;  // poison pill

        double queue_wait_ms =
            std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - item->enqueued_at).count();
        stats_.record_latency("queue_wait", queue_wait_ms);

        try {
            process(*item);
        } catch (const shared::FramingError& e) {
            LOG_ERROR(std::string("dispatcher worker: downstream write failed: ") + e.what());
            reconnect_event_.set();
        } catch (const std::exception& e) {
            LOG_ERROR(std::string("dispatcher worker: ") + e.what());
        }
    }
}

void Dispatcher::submit_response(std::map<std::string, std::string> resp, std::vector<uint8_t> raw) {
    std::string mti = resp.count("t") ? resp.at("t") : "";
    std::unique_lock<std::mutex> lock(response_queue_mutex_);
    response_queue_cv_not_full_.wait(
        lock, [this] { return response_queue_.size() < static_cast<size_t>(cfg_.queue_maxsize); });
    response_queue_.push_back(ResponseItem{std::move(resp), std::move(raw)});
    stats_.set_gauge("response_queue_depth", static_cast<int>(response_queue_.size()));
    response_queue_cv_not_empty_.notify_one();
    LOG_DEBUG("dispatcher: queued response mti=" + mti +
              " (response_queue_depth=" + std::to_string(response_queue_.size()) + ")");
}

void Dispatcher::response_worker_loop() {
    while (true) {
        std::optional<ResponseItem> item;
        {
            std::unique_lock<std::mutex> lock(response_queue_mutex_);
            response_queue_cv_not_empty_.wait(lock, [this] { return !response_queue_.empty(); });
            item = std::move(response_queue_.front());
            response_queue_.pop_front();
            stats_.set_gauge("response_queue_depth", static_cast<int>(response_queue_.size()));
            response_queue_cv_not_full_.notify_one();
        }
        if (!item.has_value()) break;  // poison pill

        try {
            handle_response(item->resp, item->raw);
        } catch (const std::exception& e) {
            LOG_ERROR(std::string("dispatcher response worker: ") + e.what());
        }
    }
}

void Dispatcher::process(RoutedMessage& msg) {
    auto& req = msg.req;
    std::string mti = req.at("t");
    std::string pan = req.count("2") ? req.at("2") : "";
    std::string upstream_stan = req.count("11") ? req.at("11") : "";

    std::string router_stan = next_stan();
    bool tracing = trace_.start(router_stan, upstream_stan, pan, mti, msg.raw);

    auto fwd = req;
    if (mti == "0100") {
        std::string f47 = req.count("47") ? req.at("47") : "";
        auto crypto_start = std::chrono::steady_clock::now();
        std::string enriched = crypto_.validate("validate_0100", pan, f47, router_stan);
        double crypto_ms =
            std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - crypto_start).count();
        stats_.record_latency("crypto_rtt", crypto_ms);
        if (tracing) {
            trace_.hop(router_stan, "crypto_call", nullptr,
                       {{"crypto_ms", std::round(crypto_ms * 1000.0) / 1000.0}, {"enriched", !enriched.empty()}});
        }
        if (!enriched.empty()) fwd["47"] = enriched;
    }
    fwd["11"] = router_stan;

    auto encoded = shared::iso_codec::encode(fwd);
    if (tracing) {
        trace_.hop(router_stan, "downstream_send", &encoded);
    }

    // msg.enqueued_at defaults to epoch only for tests that build a RoutedMessage without going
    // through submit() (which always stamps it) - fall back to now() in that case, same as
    // router_java's `msg.enqueuedAtNanos() != 0 ? ... : System.nanoTime()`.
    auto started_at = msg.enqueued_at == std::chrono::steady_clock::time_point{}
                           ? std::chrono::steady_clock::now()
                           : msg.enqueued_at;
    {
        std::lock_guard<std::mutex> lock(pending_mutex_);
        if (pending_.find(router_stan) != pending_.end()) {
            LOG_ERROR("dispatcher: router_stan collision on " + router_stan +
                      " -- overwriting a still-outstanding entry");
        }
        pending_[router_stan] =
            PendingEntry{msg.up_fd, msg.up_write_lock, upstream_stan, std::chrono::steady_clock::now(), started_at};
        stats_.set_gauge("pending_count", static_cast<int>(pending_.size()));
    }

    auto frame =
        shared::build_frame(0x00, cfg_.downstream.irm_id, cfg_.downstream.client_id, fwd.at("t"), encoded);
    downstream_.send(frame);
    stats_.record_sent();
    LOG_DEBUG("dispatcher: forwarded mti=" + mti + " to downstream, upstream_stan=" + upstream_stan +
              " router_stan=" + router_stan);
}

void Dispatcher::handle_response(const std::map<std::string, std::string>& resp, const std::vector<uint8_t>& raw) {
    std::string mti = resp.at("t");
    if (mti == "0810") return;
    if (mti != "0110" && mti != "0130" && mti != "0430") {
        LOG_WARNING("dispatcher: unexpected downstream mti=" + mti);
        return;
    }

    std::string router_stan = resp.count("11") ? resp.at("11") : "";
    trace_.hop(router_stan, "downstream_recv", &raw);
    bool tracing = trace_.is_tracing(router_stan);

    auto now = std::chrono::steady_clock::now();
    PendingEntry entry;
    {
        std::lock_guard<std::mutex> lock(pending_mutex_);
        auto it = pending_.find(router_stan);
        if (it == pending_.end()) {
            LOG_WARNING("dispatcher: no pending entry for router_stan=" + router_stan);
            if (tracing) trace_.finish(router_stan);
            return;
        }
        entry = it->second;
        pending_.erase(it);
        stats_.set_gauge("pending_count", static_cast<int>(pending_.size()));
    }
    stats_.record_latency("downstream_rtt",
                          std::chrono::duration<double, std::milli>(now - entry.created_at).count());

    auto fwd = resp;
    fwd["11"] = entry.upstream_stan;

    if (mti == "0110") {
        std::string pan = fwd.count("2") ? fwd.at("2") : "";
        std::string f47 = fwd.count("47") ? fwd.at("47") : "";
        auto crypto_start = std::chrono::steady_clock::now();
        std::string enriched = crypto_.validate("validate_0110", pan, f47, router_stan);
        double crypto_ms =
            std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - crypto_start).count();
        stats_.record_latency("crypto_rtt", crypto_ms);
        if (tracing) {
            trace_.hop(router_stan, "crypto_call", nullptr,
                       {{"crypto_ms", std::round(crypto_ms * 1000.0) / 1000.0}, {"enriched", !enriched.empty()}});
        }
        if (!enriched.empty()) fwd["47"] = enriched;
    }

    try {
        auto encoded = shared::iso_codec::encode(
            fwd, shared::iso_codec::encoding_from_config_string(cfg_.upstream.encoding));
        if (tracing) {
            trace_.hop(router_stan, "upstream_send", &encoded);
        }
        {
            std::lock_guard<std::mutex> lock(*entry.up_write_lock);
            shared::write_message(entry.up_fd, encoded, cfg_.upstream.framing);
        }
        stats_.record_sent();
        LOG_DEBUG("dispatcher: forwarded mti=" + mti + " to upstream, router_stan=" + router_stan +
                  " upstream_stan=" + entry.upstream_stan);
    } catch (const std::exception& e) {
        // Races session teardown closing the upstream socket from a different thread -- must
        // not propagate as an uncaught exception on the ds-receiver thread.
        LOG_ERROR(std::string("dispatcher: failed to write response to upstream: ") + e.what());
    }
    stats_.record_latency(
        "total", std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - entry.started_at).count());
    if (tracing) {
        trace_.finish(router_stan);
    }
}

void Dispatcher::reaper_loop() {
    while (!dispatcher_stop_.wait_for(std::chrono::seconds(1))) {
        std::vector<PendingEntry> expired;
        {
            std::lock_guard<std::mutex> lock(pending_mutex_);
            auto now = std::chrono::steady_clock::now();
            for (auto it = pending_.begin(); it != pending_.end();) {
                auto age_seconds = std::chrono::duration_cast<std::chrono::seconds>(now - it->second.created_at).count();
                if (age_seconds >= cfg_.pending_ttl_seconds) {
                    expired.push_back(it->second);
                    it = pending_.erase(it);
                } else {
                    ++it;
                }
            }
            if (!expired.empty()) {
                stats_.set_gauge("pending_count", static_cast<int>(pending_.size()));
            }
        }

        for (const auto& entry : expired) {
            try {
                std::map<std::string, std::string> decline = {{"t", "0110"}, {"11", entry.upstream_stan}, {"39", "91"}};
                auto encoded = shared::iso_codec::encode(
                    decline, shared::iso_codec::encoding_from_config_string(cfg_.upstream.encoding));
                {
                    std::lock_guard<std::mutex> lock(*entry.up_write_lock);
                    shared::write_message(entry.up_fd, encoded, cfg_.upstream.framing);
                }
                stats_.record_sent();
            } catch (const std::exception& e) {
                LOG_WARNING(std::string("dispatcher reaper: failed to send decline: ") + e.what());
            }
            LOG_WARNING("dispatcher reaper: pending entry for upstream_stan=" + entry.upstream_stan +
                        " expired, sent decline");
        }
    }
}

std::map<std::string, int> Dispatcher::purge() {
    int queue_dropped = 0;
    {
        std::lock_guard<std::mutex> lock(queue_mutex_);
        queue_dropped = static_cast<int>(queue_.size());
        queue_.clear();
        stats_.set_gauge("queue_depth", 0);
        queue_cv_not_full_.notify_all();
    }
    int response_queue_dropped = 0;
    {
        std::lock_guard<std::mutex> lock(response_queue_mutex_);
        response_queue_dropped = static_cast<int>(response_queue_.size());
        response_queue_.clear();
        stats_.set_gauge("response_queue_depth", 0);
        response_queue_cv_not_full_.notify_all();
    }
    int pending_dropped = 0;
    {
        std::lock_guard<std::mutex> lock(pending_mutex_);
        pending_dropped = static_cast<int>(pending_.size());
        pending_.clear();
        stats_.set_gauge("pending_count", 0);
    }
    return {{"queue_dropped", queue_dropped},
            {"response_queue_dropped", response_queue_dropped},
            {"pending_dropped", pending_dropped}};
}

std::vector<PendingSnapshotEntry> Dispatcher::pending_snapshot() {
    auto now = std::chrono::steady_clock::now();
    std::vector<PendingSnapshotEntry> entries;
    {
        std::lock_guard<std::mutex> lock(pending_mutex_);
        entries.reserve(pending_.size());
        for (const auto& [stan, entry] : pending_) {
            double age_seconds = std::chrono::duration<double>(now - entry.created_at).count();
            entries.push_back({stan, entry.upstream_stan, std::round(age_seconds * 1000.0) / 1000.0});
        }
    }
    std::sort(entries.begin(), entries.end(), [](const PendingSnapshotEntry& a, const PendingSnapshotEntry& b) {
        return a.age_seconds > b.age_seconds;
    });
    return entries;
}

void Dispatcher::drain_and_stop() {
    dispatcher_stop_.set();
    for (size_t i = 0; i < workers_.size(); ++i) {
        std::unique_lock<std::mutex> lock(queue_mutex_);
        queue_cv_not_full_.wait(lock, [this] { return queue_.size() < static_cast<size_t>(cfg_.queue_maxsize); });
        queue_.push_back(std::nullopt);
        queue_cv_not_empty_.notify_one();
    }
    for (size_t i = 0; i < response_workers_.size(); ++i) {
        std::unique_lock<std::mutex> lock(response_queue_mutex_);
        response_queue_cv_not_full_.wait(
            lock, [this] { return response_queue_.size() < static_cast<size_t>(cfg_.queue_maxsize); });
        response_queue_.push_back(std::nullopt);
        response_queue_cv_not_empty_.notify_one();
    }
    for (auto& t : workers_) {
        if (t.joinable()) t.join();
    }
    for (auto& t : response_workers_) {
        if (t.joinable()) t.join();
    }
    if (reaper_.joinable()) reaper_.join();

    // Session teardown (upstream or downstream disconnect, reconnect) previously discarded any
    // still-in-flight transactions with zero trace - unlike purge(), which reports
    // pending_dropped, this path left no log line explaining why a transaction just vanished.
    // There's nothing to *do* about them (the upstream client that would receive a decline is
    // already gone), but a live-diagnosis session needs the record. Safe without extra locking
    // beyond the move itself: every worker/response-worker thread that could still be mutating
    // pending_ has already been joined above.
    std::unordered_map<std::string, PendingEntry> dropped;
    {
        std::lock_guard<std::mutex> lock(pending_mutex_);
        dropped = std::move(pending_);
        pending_.clear();
    }
    for (const auto& [stan, entry] : dropped) {
        LOG_WARNING("dispatcher: session torn down with router_stan=" + stan + " still pending (upstream_stan=" +
                    entry.upstream_stan + "); abandoning");
    }
    if (!dropped.empty()) {
        LOG_WARNING("dispatcher: session teardown abandoned " + std::to_string(dropped.size()) +
                    " pending transaction(s)");
        stats_.set_gauge("pending_count", 0);
    }

    auto abandoned_traces = trace_.abandon_in_progress();
    if (!abandoned_traces.empty()) {
        LOG_WARNING("dispatcher: session teardown left " + std::to_string(abandoned_traces.size()) +
                    " in-progress trace(s) incomplete");
    }
}

}  // namespace router
